taichi/backends/opengl/codegen_opengl.cpp

```diff
@@ -616,6 +699,103 @@ class KernelGen : public IRVisitor {
     emit("}}\n");
   }
 
+  void generate_struct_for_kernel(OffloadedStmt *stmt) {
+    TI_ASSERT(stmt->task_type == OffloadedStmt::TaskType::struct_for);
+    const std::string glsl_kernel_name = make_kernel_name();
+    emit("void {}()", glsl_kernel_name);
+    this->glsl_kernel_name_ = glsl_kernel_name;
+    emit("{{ // struct for {}", stmt->snode->node_type_name);
+    {
+      ScopedIndent _s(line_appender_);
+      emit("int _tid = int(gl_GlobalInvocationID.x);");
+      emit("if (_tid >= _list_len_) return;");
+      emit("int _itv = _list_[_tid];");
+      kpa = KernelParallelAttrib(-1);
+      kpa.is_list = true;
+    }
+    stmt->body->accept(this);
+    emit("}}\n");
+  }
+
+  void generate_clear_list_kernel(OffloadedStmt *stmt) {
+    TI_ASSERT(stmt->task_type == OffloadedStmt::TaskType::clear_list);
+    const std::string glsl_kernel_name = make_kernel_name();
+    emit("void {}()", glsl_kernel_name);
+    this->glsl_kernel_name_ = glsl_kernel_name;
+    emit("{{ // clear list {}", stmt->snode->node_type_name);
+    {
+      ScopedIndent _s(line_appender_);
+      emit("_list_len_ = 0;");
+    }
+    emit("}}\n");
+  }
+
+  size_t get_snode_base_address(const SNode *snode) {
+    if (snode->type == SNodeType::root)
+      return 0;
+    int chid = find_children_id(snode);
+    const auto &parent_meta =
+        struct_compiled_->snode_map.at(snode->parent->node_type_name);
+    auto choff = parent_meta.children_offsets[chid];
+    return choff + get_snode_base_address(snode->parent);
+  }
+
+  size_t get_snode_meta_address(const SNode *snode) {
+    auto addr = get_snode_base_address(snode);
+    addr += struct_compiled_->snode_map.at(snode->node_type_name).stride;
+    addr -= opengl_get_snode_meta_size(*snode);
+    return addr;
+  }
+
+  void generate_listgen_for_dynamic(const SNode *snode) {
+    TI_ASSERT(snode->type == SNodeType::dynamic);
+    // the `length` field of a dynamic SNode is at it's end:
+    // | x[0] | x[1] | x[2] | x[3] | ... | len |
+    TI_ASSERT_INFO(snode->parent->type == SNodeType::root,
+                   "Non-top-level dynamic not supported yet on OpenGL");
+    size_t addr = get_snode_meta_address(snode);
+    emit("_list_len_ = _data_i32_[{} >> 2];", addr);
+    emit("for (int i = 0; i < _list_len_; i++) {{");
+    {
+      ScopedIndent _s(line_appender_);
+      emit("_list_[i] = i;");
+    }
+    emit("}}");
+  }
+
+  void generate_listgen_for_dense(const SNode *snode) {
```

Comment:
It feels a bit strange: if we use this on a dense with more than one dimension, it probably won't work for now. But struct-for on a dense-only hierarchy is demoted to range-for, so it doesn't matter..

Reply:
Not sure how to support multi-dim listgen now, will do iapr.